src/nbla/cuda/solver/generic/clip_grad.cuh

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) 2017 Sony Corporation. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NBLA_CUDA_FUNCTION_GENERIC_CLIP_GRAD_CUH
+#define NBLA_CUDA_FUNCTION_GENERIC_CLIP_GRAD_CUH
+
+#include <nbla/cuda/common.hpp>
+#include <nbla/function/pow_scalar.hpp>
+#include <nbla/function/sum.hpp>
+#include <nbla/variable.hpp>
+
+namespace nbla {
+template <typename T>
+__global__ void kernel_clip_grad_by_norm(const int num, T *grad, const T norm,
+                                         const float clip_norm) {
+  NBLA_CUDA_KERNEL_LOOP(idx, num) {
+    grad[idx] = clip_norm * grad[idx] / norm;
+  }
+}
+
+template <typename T>
+void clip_grad_by_norm_cuda(const Context &ctx,
+                            const shared_ptr<Variable> param, float clip_norm) {
+  cuda_set_device(std::stoi(ctx.device_id));
+
+  Variable g(param->grad());
+  Variable g_pow(param->shape());
+  Variable sum(Shape_t{});
+
+  // calculate g^2
+  auto f_pow_scalar = create_PowScalar(ctx, 2.0);
+  f_pow_scalar->setup(Variables{&g}, Variables{&g_pow});
+  f_pow_scalar->forward(Variables{&g}, Variables{&g_pow});
+
+  // sum all gradients
+  vector<int> axis;
+  for (int i = 0; i < param->ndim(); ++i)
+    axis.push_back(i);
+  auto f_sum = create_Sum(ctx, axis, false);
+  f_sum->setup(Variables{&g_pow}, Variables{&sum});
+  f_sum->forward(Variables{&g_pow}, Variables{&sum});
+
+  // check norm on host to reduce unnecessary kernel launch
+  const T *l2sum = sum.get_data_pointer<T>(ctx);
+  T l2sum_cpu;
+  cudaMemcpy(&l2sum_cpu, l2sum, sizeof(T), cudaMemcpyDeviceToHost);
+  T norm = sqrt(l2sum_cpu);
+  if (norm > clip_norm) {
```

Comment:
You should evaluate `norm > clip_norm` on GPU (in cuda kernel) to avoid unnecessary synchronization between cpu and gpu.
Current implementation stops cpu execution at this `cudaMemcpy` until all gpu executions queued before are finished.

Reply:
Just pass `clip_norm` and `norm` to `kernel_clip_grad_by_norm`, and evaluate `evaluate norm > clip_norm` in the kernel.